standard/types.md

```diff
@@ -705,3 +714,182 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+> string s2 = null;     // OK; warnings are disabled
+> s2 = null;            // OK; warnings are disabled
+> char c1 = s2[1];      // OK; warnings are disabled; throws NullReferenceException
+> c1 = s2![1];          // No warnings
+> ```
+>
+> *end example*
+
+#### §Nullable-Warning-Context Nullable warnings
+
+When the nullable context is ***warnings***:
+
+The compiler uses static flow analysis to determine the null state of any reference variable. When nullable warnings are enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null*, *maybe null*, or *maybe default* and
+
+- A warning shall be generated if a reference variable that has been determined to be *maybe null*, is dereferenced.
+- The state of a reference expression is *maybe null* unless the compiler can determine one of two conditions:
+  1. The variable has been definitely assigned a non-`null` value.
+  1. The variable has been checked against `null` before dereferencing it.
+
+Some warnings are generated when the nullable context is set to *warnings*. (Examples include throwing a value that may be `null`, and dereferencing a possibly null reference.) Other warnings are generated only if the nullable context is enabled (§Nullable-Warning-Context). (Examples include converting a null literal or possible null value to a non-nullable type, possible null reference assignment, and possible null reference return.)
+
+> *Example*:
+>
+> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext3", ignoredWarnings:["CS0219"], expectedWarnings:["CS8632", "CS8602"], expectedException:"NullReferenceException"} -->
+> ```csharp
+> #nullable disable annotations
+> #nullable enable warnings
+> string? s1 = null;    // OK; ? makes s2 nullable
+> string s2 = null;     // OK; null-state of s2 is "maybe null"
+> s2 = null;            // OK; null-state of s2 is "maybe null"
+> char c1 = s2[1];      // Warning; dereference of a possible null; throws NullReferenceException
+> c1 = s2![1];          // The warning is suppressed
+> ```
+>
+> *end example*
+
+#### §Nullable-enable-Context Nullable enable
+
+When the nullable context is ***enabled***:
+
+- For any reference type `T`, the annotation `?` in `T?` makes `T?` a nullable type, whereas the unannotated `T` is non-nullable.
+- The compiler uses static flow analysis to determine the null state of any reference variable. When nullable warnings are enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null*, *maybe null*, or *maybe default* and
+  - A warning shall be generated when a possible null value is assigned to a variable of a non-nullable type.
+  - A warning shall be generated if a reference expression that has been determined to be *maybe null*, is dereferenced.
+  - The state of a non-nullable reference variable is *not null* unless the compiler can determine that the variable was assigned to an expression that is *maybe null*.
```

Comment:
```suggestion
  - The state of a non-nullable reference variable is *not null* unless the compiler can determine that the variable was assigned an expression that is *maybe null*.
```

Reply:
(Or "assigned with the result of an expression" or even "assigned with the result of evaluating an expression" - we should look at what terminology we use elsewhere for assignments.)